src/number_theory/modular.lean

```diff
@@ -363,6 +373,267 @@ begin
+  { simp [S, matrix.mul_apply, fin.sum_univ_succ], },
+end
+
+/-- Nontrivial lemma: if `|x|<1/2` and `n:ℤ`, then `2nx+n^2≥0`. (False for `n:ℝ`!) -/
+lemma _root_.int.non_neg_of_lt_half (n : ℤ) (x : ℝ) (hx : |x| < 1/2) : (0:ℝ) ≤ 2 * n * x + n * n :=
+begin
+  rw abs_lt at hx,
+  have : (0:ℝ) ≤ n*n := by nlinarith,
+  cases n,
+  { -- n ≥ 0
+    have : (n:ℝ) = (int.of_nat n) := by simp,
+    have : (0:ℝ) ≤ n := by simp,
+    cases lt_or_ge x 0,
+    {  -- x < 0
+      cases n,
+      { simp, },
+      { -- n ≥ 1
+        have eq1 : (1:ℝ) ≤ int.of_nat n.succ := by simp,
+        have eq2 : (1:ℝ) ≤ (int.of_nat n.succ) * (int.of_nat n.succ) := by nlinarith,
+        have eq3 : (-1:ℝ) ≤ 2 * x := by nlinarith,
+        have eq4 : (0:ℝ) ≤ 2 * x + int.of_nat n.succ := by linarith,
+        have eq5 : (0:ℝ) ≤ (2 * x + int.of_nat n.succ)*(int.of_nat n.succ) := by nlinarith,
+        convert eq5 using 1,
+        ring, }, },
+    { -- x ≥ 0
+      have : (0:ℝ) ≤ 2*n*x := by nlinarith,
+      nlinarith, }, },
+  { -- n ≤ -1
+    have := int.neg_succ_of_nat_coe n,
+    set k := int.neg_succ_of_nat n,
+    have eq1 : k ≤ -1,
+    { have : 1 ≤ 1 + n := by simp,
+      have :  -((1:ℤ) + n) ≤ -1,
+      { have : 0 ≤ n := by simp,
+        linarith, },
+      convert this using 1,
+      simp [this_1],
+      ring, },
+    have eq1' : (k:ℝ) ≤ -1 := by exact_mod_cast eq1,
+    cases lt_or_ge x 0,
+    { -- x < 0
+      have : (0:ℝ) ≤ 2*k*x := by nlinarith,
+      have eq2 : 1 ≤ k*k  := by nlinarith,
+      linarith, },
+    { -- x ≥ 0
+      have eq2 : (2:ℝ) * x + k ≤ 0 := by nlinarith,
+      nlinarith, }, },
+end
+
+/-- If `z∈𝒟ᵒ`, and `n:ℤ`, then `|z+n|>1`.   -/
+lemma move_by_T {z : ℍ} (hz : z ∈ 𝒟ᵒ) (n : ℤ) : 1 < norm_sq (((T_pow n) • z) : ℍ) :=
+begin
+  rw T_pow,
+  simp,
+  rw complex.norm_sq_apply,
+  have hz1 : 1 < z.re * z.re + z.im * z.im,
+  { have := hz.1,
+    rw norm_sq at this,
+    convert this using 1, },
+  rw (by simp : ((z:ℂ) + n).im = z.im),
+  rw (by simp : ((z:ℂ) + n).re = z.re + n),
+  rw (by ring : (z.re + ↑n) * (z.re + ↑n) = z.re * z.re + 2 * n * z.re + n * n),
+  have : 0 ≤  2 * ↑n * z.re + ↑n * ↑n := int.non_neg_of_lt_half n (z.re) hz.2,
+  convert add_lt_add_of_le_of_lt this hz1 using 1,
+  { simp, },
+  { ring_nf, },
+end
+
+/-- If `c=1`, then `[[1,-a],[0,1]]*g = S * [[1,d],[0,1]]`. -/
+lemma T_pow_S_of_g (g : SL(2,ℤ)) (hc : g 1 0 = 1) : T_pow (- g 0 0) * g = S * T_pow (g 1 1) :=
+begin
+  rw g_is_of_c_is_one g hc,
+  ext i,
+  fin_cases i; fin_cases j,
+  { simp [T_pow, S, matrix.mul_apply, fin.sum_univ_succ], },
+  { simp [T_pow, S, matrix.mul_apply, fin.sum_univ_succ],
+    ring },
+  { simp [T_pow, S, matrix.mul_apply, fin.sum_univ_succ], },
+  { simp [T_pow, S, matrix.mul_apply, fin.sum_univ_succ], },
+end
+
+/-- If both `z` and `g•z` are in `𝒟ᵒ`, then `c` can't be `1`. -/
+lemma false_of_c_eq_one {z : ℍ} {g : SL(2,ℤ)} (hc : g 1 0 = 1) (hz : z ∈ 𝒟ᵒ) (hg : g • z ∈ 𝒟ᵒ) :
```

Comment:
Instead of having `: false` it seems more reasonable to have `: g 1 0 \neq 1` (moreover this is what the description says).